src/Nethermind/Nethermind.Synchronization/Blocks/BlockDownloader.cs

```diff
@@ -175,22 +165,287 @@ private async Task<long> DownloadBlocks(PeerInfo? bestPeer, BlocksRequest blocks
                     }
                 }
 
+                // Should only happen because of a lot of reorg
+                // or if the `HeaderLookupSize` become smaller significantly.
+                if (_downloadRequests.Count > headers.Count * 2)
+                {
+                    PruneRequestMap(headers);
+                }
+
                 if (blocksSynced > 0)
                 {
                     _syncReport.FullSyncBlocksDownloaded.Update(_blockTree.BestSuggestedHeader?.Number ?? 0);
                 }
-                else
+
+                _syncReport.FullSyncBlocksDownloaded.CurrentQueued = _downloadRequests.Count;
+
+                if (satisfiedEntry.Count == 0) // Nothing left to process
+                {
+                    return await AssembleRequest(headers, downloadReceipts, cancellation);
+                }
+            }
+        }
+
+        public void PruneDownloadBuffer()
+        {
+            _downloadRequests.Clear();
+        }
+
+        private void PruneRequestMap(IOwnedReadOnlyList<BlockHeader> currentHeaders)
+        {
+            HashSet<Hash256> currentHeaderHashes = currentHeaders.Select(h => h.Hash).ToHashSet();
+            HashSet<Hash256> toRemove = new HashSet<Hash256>();
+            foreach (var kv in _downloadRequests)
+            {
+                if (!currentHeaderHashes.Contains(kv.Key))
+                {
+                    toRemove.Add(kv.Key);
```

Comment:
To be honest, you don't need a `toRemove` collection at all. `ConcurrentDictionary` supports iterating and removing in performant way. It iterates on a snapshot-like (but light) enumerator.

```suggestion
                    _downloadRequests.Remove(kv.Key, out _);
```